# Eigenvalues and eigenvectors

The goal of this notebook is to show how to compute eigenvalues and eigenvectors of a square matrix using numpy.

In [1]:
import numpy as np

Here we define a sample matrix:

In [2]:
A = np.array([[8, 2, 5, 8],
              [2, 9, 8, 5],
              [5, 8, 8, 9],
              [8, 5, 9, 8]])

The function `np.linalg.eig` returns a tuple giving eigenvalues and eigenvectors of a matrix:

In [3]:
eigenvals, eigenvects = np.linalg.eig(A)

Eigenvalues of the matrix `A`:

In [4]:
print(eigenvals)

[27.15475075  7.567203    0.33785338 -2.05980714]


`eigenvects` is as matrix of eigenvectors of `A` with each column giving one eigenvector:

In [5]:
print(eigenvects)

[[ 0.42453415  0.64417668  0.5774985   0.26702554]
 [ 0.44593734 -0.68865353  0.53436729 -0.2033416 ]
 [ 0.55832431 -0.202712   -0.48210397  0.64401676]
 [ 0.55604371  0.26400866 -0.38538662 -0.68745325]]


The first column of is an eigenvector corresponding to the first eigenvalue, the seconds columns is an eigenvector corresponding to the second eigenvalue etc. To verify, lets multiply the first column of `eigenvects` by the matrix A:

In [6]:
A@eigenvects[:, 0]

array([11.52811913, 12.10931739, 15.16115735, 15.09922838])

Notice the result is the same as multiplying this columns by the first eigenvalue: 

In [7]:
eigenvals[0]*eigenvects[:, 0]

array([11.52811913, 12.10931739, 15.16115735, 15.09922838])

**Note** All eigenvectors returned by `np.linalg.eig` are vectors of length 1. 

## Scientific notation

If eigenvalues or eigenvectors computed by `np.linalg.eig` involve small or large numbers, numpy will display them using the scientific notation. For example, 
here are eigenvales of the matrix `0.0001*A`:

In [8]:
eigenvals, eigenvects = np.linalg.eig(0.0001*A)
print(eigenvals)

[ 2.71547508e-03  7.56720300e-04  3.37853384e-05 -2.05980714e-04]


The expression `2.71547508e-03` represents the number  $2.71547508\cdot 10^{-3} = 0.00271547508$. In the same way, `2.71547508e+03` would represent the number  $2.71547508\cdot 10^{3} = 2715.47508$.

In [9]:
L = np.array([[ 3,  0,  0, -1,  0,  0, -1,  0, -1,  0],
              [ 0,  3,  0, -1, -1,  0,  0, -1,  0,  0],
              [ 0,  0,  3,  0, -1, -1,  0,  0,  0, -1],
              [-1, -1,  0,  4,  0,  0, -1,  0, -1,  0],
              [ 0, -1, -1,  0,  4, -1,  0, -1,  0,  0],
              [ 0,  0, -1,  0, -1,  4,  0, -1,  0, -1],
              [-1,  0,  0, -1,  0,  0,  2,  0,  0,  0],
              [ 0, -1,  0,  0, -1, -1,  0,  4,  0, -1],
              [-1,  0,  0, -1,  0,  0,  0,  0,  2,  0],
              [ 0,  0, -1,  0,  0, -1,  0, -1,  0,  3]])

In [10]:
eigenvals, eigenvects = np.linalg.eig(L)

In [11]:
print(eigenvals[1])

0.24846792845247026


In [12]:
print(eigenvects[:,1])

[-0.40240673  0.072676    0.31607615 -0.30242156  0.25119595  0.30242156
 -0.40240673  0.25119595 -0.40240673  0.31607615]


In [13]:
second_v = np.sort(eigenvects[:,1])
second_v

array([-0.40240673, -0.40240673, -0.40240673, -0.30242156,  0.072676  ,
        0.25119595,  0.25119595,  0.30242156,  0.31607615,  0.31607615])

### K = $4$

In [14]:
sum_x_jk = sum(-second_v[0:4])
sum_x_jk

1.5096417488627245

In [15]:
sum_x_ik = sum(second_v[6:])
sum_x_ik

1.1857698057657784

In [16]:
sum_x_ik < sum_x_jk

True

### K = $6$

In [17]:
sum_x_jk = sum(-second_v[0:6])
sum_x_jk

1.1857698057657773

In [18]:
sum_x_ik = sum(second_v[4:])
sum_x_ik

1.5096417488627258

In [19]:
sum_x_ik > sum_x_jk

True

In [20]:
second_v[0:4]

array([-0.40240673, -0.40240673, -0.40240673, -0.30242156])

In [21]:
second_v[6:]

array([0.25119595, 0.30242156, 0.31607615, 0.31607615])

In [22]:
A = [[18,3,11,-2],
     [3,18,-2,11],
     [11,-2,18,3],
     [-2,11,3,18]]

In [23]:
eigenvals, eigenvects = np.linalg.eig(A)

In [24]:
eigenvals

array([ 2., 12., 30., 28.])